Please consider 32bit support, if possible.

We have dissec.cstruct packaged for Kali, and in our build testing, 32bit builds all fail.  If this is expected and we should restrict the package to 64bit, that is fine, please let us know!

The tests that fail look like this:

```
I: pybuild base:240: cd /builds/kalilinux/packages/dissect.cstruct/debian/output/source_dir/.pybuild/cpython3_3.11_dissect.cstruct/build; python3.11 -m pytest tests
============================= test session starts ==============================
platform linux -- Python 3.11.4, pytest-7.4.0, pluggy-1.2.0
rootdir: /builds/kalilinux/packages/dissect.cstruct/debian/output/source_dir/.pybuild/cpython3_3.11_dissect.cstruct/build
collected 178 items
tests/test_align.py .......F                                             [  4%]
tests/test_basic.py .....................................                [ 25%]
tests/test_bitbuffer.py .                                                [ 25%]
tests/test_bitfield.py ..................                                [ 35%]
tests/test_bytesinteger.py .............                                 [ 43%]
tests/test_ctypes_type.py .....                                          [ 46%]
tests/test_enum.py .......                                               [ 50%]
tests/test_expression.py ..............................................  [ 75%]
tests/test_flag.py ...                                                   [ 77%]
tests/test_packedtype.py ......                                          [ 80%]
tests/test_pointer.py FFFFFF.                                            [ 84%]
tests/test_struct.py .................                                   [ 94%]
tests/test_union.py .....                                                [ 97%]
tests/test_util.py ...                                                   [ 98%]
tests/test_utils.py ..                                                   [100%]
=================================== FAILURES ===================================
______________________________ test_align_pointer ______________________________
    def test_align_pointer():
        d = """
        struct test {
            uint32  a;
            uint32  *b;
            uint16  c;
            uint16  d;
        };
        """
        c = cstruct.cstruct()
        c.load(d, align=True)
    
        fields = c.test.fields
        assert c.test.align
>       assert c.test.alignment == 8
E       assert 4 == 8
E        +  where 4 = <Structure test>.alignment
E        +    where <Structure test> = <dissect.cstruct.cstruct.cstruct object at 0xf7181650>.test
tests/test_align.py:294: AssertionError
___________________________ test_pointer_basic[True] ___________________________
compiled = True
    @pytest.mark.parametrize("compiled", [True, False])
    def test_pointer_basic(compiled):
        cdef = """
        struct ptrtest {
            uint32  *ptr1;
            uint32  *ptr2;
        };
        """
        cs = cstruct.cstruct(pointer="uint16")
        cs.load(cdef, compiled=compiled)
    
        assert verify_compiled(cs.ptrtest, compiled)
    
        buf = b"\x04\x00\x08\x00\x01\x02\x03\x04\x05\x06\x07\x08"
        obj = cs.ptrtest(buf)
    
        assert obj.ptr1 != 0
        assert obj.ptr2 != 0
        assert obj.ptr1 != obj.ptr2
>       assert obj.ptr1 == 4
E       assert <Pointer uint32 @ 0x80004> == 4
E        +  where <Pointer uint32 @ 0x80004> = <ptrtest ptr1=<Pointer uint32 @ 0x80004>, ptr2=<Pointer uint32 @ 0x4030201>>.ptr1
tests/test_pointer.py:28: AssertionError
__________________________ test_pointer_basic[False] ___________________________
compiled = False
    @pytest.mark.parametrize("compiled", [True, False])
    def test_pointer_basic(compiled):
        cdef = """
        struct ptrtest {
            uint32  *ptr1;
            uint32  *ptr2;
        };
        """
        cs = cstruct.cstruct(pointer="uint16")
        cs.load(cdef, compiled=compiled)
    
        assert verify_compiled(cs.ptrtest, compiled)
    
        buf = b"\x04\x00\x08\x00\x01\x02\x03\x04\x05\x06\x07\x08"
        obj = cs.ptrtest(buf)
    
        assert obj.ptr1 != 0
        assert obj.ptr2 != 0
        assert obj.ptr1 != obj.ptr2
>       assert obj.ptr1 == 4
E       assert <Pointer uint32 @ 0x80004> == 4
E        +  where <Pointer uint32 @ 0x80004> = <ptrtest ptr1=<Pointer uint32 @ 0x80004>, ptr2=<Pointer uint32 @ 0x4030201>>.ptr1
tests/test_pointer.py:28: AssertionError
__________________________ test_pointer_struct[True] ___________________________
compiled = True
    @pytest.mark.parametrize("compiled", [True, False])
    def test_pointer_struct(compiled):
        cdef = """
        struct test {
            char    magic[4];
            wchar   wmagic[4];
            uint8   a;
            uint16  b;
            uint32  c;
            char    string[];
            wchar   wstring[];
        };
    
        struct ptrtest {
            test    *ptr;
        };
        """
        cs = cstruct.cstruct(pointer="uint16")
        cs.load(cdef, compiled=compiled)
    
        assert verify_compiled(cs.test, compiled)
        assert verify_compiled(cs.ptrtest, compiled)
    
        buf = b"\x02\x00testt\x00e\x00s\x00t\x00\x01\x02\x03\x04\x05\x06\x07lalala\x00t\x00e\x00s\x00t\x00\x00\x00"
        obj = cs.ptrtest(buf)
    
        assert obj.ptr != 0
    
>       assert obj.ptr.magic == b"test"
tests/test_pointer.py:72: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
dissect/cstruct/types/pointer.py:55: in __getattr__
    return getattr(self.dereference(), attr)
dissect/cstruct/types/pointer.py:122: in dereference
    value = self._type._read(self._stream, self._ctx)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
self = <Structure test +compiled>, stream = <_io.BytesIO object at 0xf788e410>
context = OrderedDict([('ptr', <Pointer <Structure test +compiled> @ 0x65740002>)])
>   ???
E   EOFError
<compiled test>:15: EOFError
__________________________ test_pointer_struct[False] __________________________
compiled = False
    @pytest.mark.parametrize("compiled", [True, False])
    def test_pointer_struct(compiled):
        cdef = """
        struct test {
            char    magic[4];
            wchar   wmagic[4];
            uint8   a;
            uint16  b;
            uint32  c;
            char    string[];
            wchar   wstring[];
        };
    
        struct ptrtest {
            test    *ptr;
        };
        """
        cs = cstruct.cstruct(pointer="uint16")
        cs.load(cdef, compiled=compiled)
    
        assert verify_compiled(cs.test, compiled)
        assert verify_compiled(cs.ptrtest, compiled)
    
        buf = b"\x02\x00testt\x00e\x00s\x00t\x00\x01\x02\x03\x04\x05\x06\x07lalala\x00t\x00e\x00s\x00t\x00\x00\x00"
        obj = cs.ptrtest(buf)
    
        assert obj.ptr != 0
    
>       assert obj.ptr.magic == b"test"
tests/test_pointer.py:72: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
dissect/cstruct/types/pointer.py:55: in __getattr__
    return getattr(self.dereference(), attr)
dissect/cstruct/types/pointer.py:122: in dereference
    value = self._type._read(self._stream, self._ctx)
dissect/cstruct/types/structure.py:188: in _read
    value = field_type._read(stream, result)
dissect/cstruct/types/packedtype.py:20: in _read
    return self._read_array(stream, 1, context)[0]
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
self = uint8, stream = <_io.BytesIO object at 0xf71c2e10>, count = 1
context = OrderedDict([('magic', b''), ('wmagic', '')])
    def _read_array(self, stream: BinaryIO, count: int, context: dict[str, Any] = None) -> List[int]:
        length = self.size * count
        data = stream.read(length)
        fmt = self.cstruct.endian + str(count) + self.packchar
    
        if len(data) != length:
>           raise EOFError(f"Read {len(data)} bytes, but expected {length}")
E           EOFError: Read 0 bytes, but expected 1
dissect/cstruct/types/packedtype.py:28: EOFError
_________________________ test_array_of_pointers[True] _________________________
compiled = True
    @pytest.mark.parametrize("compiled", [True, False])
    def test_array_of_pointers(compiled):
        cdef = """
        struct mainargs {
            uint8_t argc;
            char *args[4];
        }
        """
        cs = cstruct.cstruct(pointer="uint16")
        cs.load(cdef, compiled=compiled)
    
        assert verify_compiled(cs.mainargs, compiled)
    
        buf = b"\x02\x09\x00\x16\x00\x00\x00\x00\x00argument one\x00argument two\x00"
        obj = cs.mainargs(buf)
    
        assert obj.argc == 2
>       assert obj.args[2] == 0
E       assert <Pointer char @ 0x75677261> == 0
tests/test_pointer.py:103: AssertionError
________________________ test_array_of_pointers[False] _________________________
compiled = False
    @pytest.mark.parametrize("compiled", [True, False])
    def test_array_of_pointers(compiled):
        cdef = """
        struct mainargs {
            uint8_t argc;
            char *args[4];
        }
        """
        cs = cstruct.cstruct(pointer="uint16")
        cs.load(cdef, compiled=compiled)
    
        assert verify_compiled(cs.mainargs, compiled)
    
        buf = b"\x02\x09\x00\x16\x00\x00\x00\x00\x00argument one\x00argument two\x00"
        obj = cs.mainargs(buf)
    
        assert obj.argc == 2
>       assert obj.args[2] == 0
E       assert <Pointer char @ 0x75677261> == 0
tests/test_pointer.py:103: AssertionError
=========================== short test summary info ============================
FAILED tests/test_align.py::test_align_pointer - assert 4 == 8
 +  where 4 = <Structure test>.alignment
 +    where <Structure test> = <dissect.cstruct.cstruct.cstruct object at 0xf7181650>.test
FAILED tests/test_pointer.py::test_pointer_basic[True] - assert <Pointer uint32 @ 0x80004> == 4
 +  where <Pointer uint32 @ 0x80004> = <ptrtest ptr1=<Pointer uint32 @ 0x80004>, ptr2=<Pointer uint32 @ 0x4030201>>.ptr1
FAILED tests/test_pointer.py::test_pointer_basic[False] - assert <Pointer uint32 @ 0x80004> == 4
 +  where <Pointer uint32 @ 0x80004> = <ptrtest ptr1=<Pointer uint32 @ 0x80004>, ptr2=<Pointer uint32 @ 0x4030201>>.ptr1
FAILED tests/test_pointer.py::test_pointer_struct[True] - EOFError
FAILED tests/test_pointer.py::test_pointer_struct[False] - EOFError: Read 0 bytes, but expected 1
FAILED tests/test_pointer.py::test_array_of_pointers[True] - assert <Pointer char @ 0x75677261> == 0
FAILED tests/test_pointer.py::test_array_of_pointers[False] - assert <Pointer char @ 0x75677261> == 0
======================== 7 failed, 171 passed in 0.68s =========================
```

You can see our jobs for [armel](https://gitlab.com/kalilinux/packages/dissect.cstruct/-/jobs/4862899223), [armhf](https://gitlab.com/kalilinux/packages/dissect.cstruct/-/jobs/4862899228) and [i386](https://gitlab.com/kalilinux/packages/dissect.cstruct/-/jobs/4862899206)

While we are still on 3.6 in the packaging, my local testing for 3.9 still shows the same test failures.

--- a/dissect/cstruct/cstruct.py
+++ b/dissect/cstruct/cstruct.py
@@ -2,7 +2,7 @@ from __future__ import print_function
 
 import ctypes as _ctypes
 import sys
-from typing import Any, BinaryIO
+from typing import Any, BinaryIO, Optional
 
 from dissect.cstruct.exceptions import ResolveError
 from dissect.cstruct.parser import CStyleParser, TokenParser
@@ -30,7 +30,7 @@ class cstruct:
     DEF_CSTYLE = 1
     DEF_LEGACY = 2
 
-    def __init__(self, endian: str = "<", pointer: str = None):
+    def __init__(self, endian: str = "<", pointer: Optional[str] = None):
         self.endian = endian
 
         self.consts = {}
@@ -146,7 +146,7 @@ class cstruct:
         }
         # fmt: on
 
-        pointer = pointer or "uint64" if sys.maxsize > 2**32 else "uint32"
+        pointer = pointer or ("uint64" if sys.maxsize > 2**32 else "uint32")
         self.pointer = self.resolve(pointer)
         self._anonymous_count = 0
 

--- a/tests/test_align.py
+++ b/tests/test_align.py
@@ -289,7 +289,7 @@ def test_align_pointer():
     c = cstruct.cstruct(pointer="uint64")
     c.load(d, align=True)
 
-    assert c.pointer == "uint64"
+    assert c.pointer is c.uint64
 
     fields = c.test.fields
     assert c.test.align

--- a/tests/test_align.py
+++ b/tests/test_align.py
@@ -286,8 +286,10 @@ def test_align_pointer():
         uint16  d;
     };
     """
-    c = cstruct.cstruct()
+    c = cstruct.cstruct(pointer="uint64")
     c.load(d, align=True)
+
+    assert c.pointer == "uint64"
 
     fields = c.test.fields
     assert c.test.align

--- a/tests/test_pointer.py
+++ b/tests/test_pointer.py
@@ -1,5 +1,6 @@
-import pytest
 from unittest.mock import patch
+
+import pytest
 from dissect import cstruct
 
 from dissect.cstruct.types.pointer import PointerInstance

--- a/tests/test_pointer.py
+++ b/tests/test_pointer.py
@@ -18,7 +18,7 @@ def test_pointer_basic(compiled):
     cs.load(cdef, compiled=compiled)
 
     assert verify_compiled(cs.ptrtest, compiled)
-    assert cs.pointer == "uint16"
+    assert cs.pointer is cs.uint16
 
     buf = b"\x04\x00\x08\x00\x01\x02\x03\x04\x05\x06\x07\x08"
     obj = cs.ptrtest(buf)
@@ -64,7 +64,7 @@ def test_pointer_struct(compiled):
 
     assert verify_compiled(cs.test, compiled)
     assert verify_compiled(cs.ptrtest, compiled)
-    assert cs.pointer == "uint16"
+    assert cs.pointer is cs.uint16
 
     buf = b"\x02\x00testt\x00e\x00s\x00t\x00\x01\x02\x03\x04\x05\x06\x07lalala\x00t\x00e\x00s\x00t\x00\x00\x00"
     obj = cs.ptrtest(buf)
@@ -97,7 +97,7 @@ def test_array_of_pointers(compiled):
     cs.load(cdef, compiled=compiled)
 
     assert verify_compiled(cs.mainargs, compiled)
-    assert cs.pointer == "uint16"
+    assert cs.pointer is cs.uint16
 
     buf = b"\x02\x09\x00\x16\x00\x00\x00\x00\x00argument one\x00argument two\x00"
     obj = cs.mainargs(buf)

--- a/tests/test_pointer.py
+++ b/tests/test_pointer.py
@@ -18,6 +18,7 @@ def test_pointer_basic(compiled):
     cs.load(cdef, compiled=compiled)
 
     assert verify_compiled(cs.ptrtest, compiled)
+    assert cs.pointer == "uint16"
 
     buf = b"\x04\x00\x08\x00\x01\x02\x03\x04\x05\x06\x07\x08"
     obj = cs.ptrtest(buf)
@@ -63,6 +64,7 @@ def test_pointer_struct(compiled):
 
     assert verify_compiled(cs.test, compiled)
     assert verify_compiled(cs.ptrtest, compiled)
+    assert cs.pointer == "uint16"
 
     buf = b"\x02\x00testt\x00e\x00s\x00t\x00\x01\x02\x03\x04\x05\x06\x07lalala\x00t\x00e\x00s\x00t\x00\x00\x00"
     obj = cs.ptrtest(buf)
@@ -95,6 +97,7 @@ def test_array_of_pointers(compiled):
     cs.load(cdef, compiled=compiled)
 
     assert verify_compiled(cs.mainargs, compiled)
+    assert cs.pointer == "uint16"
 
     buf = b"\x02\x09\x00\x16\x00\x00\x00\x00\x00argument one\x00argument two\x00"
     obj = cs.mainargs(buf)

--- a/tests/test_pointer.py
+++ b/tests/test_pointer.py
@@ -1,4 +1,5 @@
 import pytest
+from unittest.mock import patch
 from dissect import cstruct
 
 from dissect.cstruct.types.pointer import PointerInstance
@@ -145,3 +146,16 @@ def test_pointer_arithmetic():
 
     inst |= 8
     assert inst._addr == 12
+
+
+def test_pointer_sys_size():
+    with patch("sys.maxsize", 2**64):
+        c = cstruct.cstruct()
+        assert c.pointer is c.uint64
+
+    with patch("sys.maxsize", 2**32):
+        c = cstruct.cstruct()
+        assert c.pointer is c.uint32
+
+    c = cstruct.cstruct(pointer="uint16")
+    assert c.pointer is c.uint16

